### DISCUSSION TITLE
Permission denied when running in nektos/act

### DISCUSSION THREAD

**Discussion by surgiie** [stefanzweifel/git-auto-commit-action#312] [Help] [answered]
I'm using https://github.com/nektos/act to test my github actions locally, but im getting permission denied errors when running this action:

```
[Deployment/deployment]   💬  ::debug::Apply status options                                                                                                                                    
| INPUT_FILE_PATTERN: *.yaml                                                                                                                                                                   
| INPUT_BRANCH value:                                                                                                                                                                          
| git@github.com: Permission denied (publickey).                                                                                                                                               
| fatal: Could not read from remote repository.                                                                                                                                                
|                                                                                                                                                                                              
| Please make sure you have the correct access rights                                                                                                                                          
| and the repository exists.                                                                                                                                                                   
| Error: Invalid status code: 128                                                                                                                                                              
|     at ChildProcess.<anonymous> (/run/act/actions/stefanzweifel-git-auto-commit-action@v5/index.js:17:19)                                                                                    
|     at ChildProcess.emit (node:events:514:28)                                                                                                                                                
|     at maybeClose (node:internal/child_process:1105:16)                                                                                                                                      
|     at ChildProcess._handle.onexit (node:internal/child_process:305:5) {                                                                                                                     
|   code: 128                                                                                                                                                                                  
| }                                                                                                                                                                                            
| Error: Invalid status code: 128                                                                                                                                                              
|     at ChildProcess.<anonymous> (/run/act/actions/stefanzweifel-git-auto-commit-action@v5/index.js:17:19)                                                                                    
|     at ChildProcess.emit (node:events:514:28)                                                                                                                                                
|     at maybeClose (node:internal/child_process:1105:16)                                                                                                                                      
|     at ChildProcess._handle.onexit (node:internal/child_process:305:5)      
```

I'm using a token to checkout the repo ( this token belongs to me when running locally, it definitely has permissions needed for this)

```yaml
- uses: actions/checkout@master
  with:
    fetch-depth: 2
    token: ${{ secrets.GH_TOKEN }}
```
I have a step to add file and and commit it 

```yaml

- name: Add test file
  run: |
      # ... 
      touch test.yaml
      
 -  name: Commit test.yaml
    uses: stefanzweifel/git-auto-commit-action@v5
    with:
        commit_message: Commit test.yaml [skip actions]
        branch: ${{ github.head_ref }}
        file_pattern: '*.yaml'
```

When i push to the repo and the action workflow runs on github, it succeeds so im unsure of what im missing here?

Anything stand out or am I missing something? Im an administrator on the repo in question and my token definitely has permissions so im unsure of why this is failing.

**Comment by stefanzweifel** [comment]
I haven't used the action with `act` before, so I can't give an immediate answer.

That you use `actions/checkout@master` and not a pinned version like `actions/checkout@v4` probably shouldn't be an issue here.

Maybe this is related to [`git fetch`](https://github.com/stefanzweifel/git-auto-commit-action/blob/master/entrypoint.sh#L96), as the error happens near that line in the script and you use a custom fetch-depth.
Could you try and disable that by using `skip_fetch: true`?

As the Action works as expected when run on GitHub Actions and not in Act, I wonder if this is related to how Act communicated with the "outside". 
When using Act, does one usually expect that the commit are actually pushed to GitHub. Or is Act more treated as a sandbox environment where nothing can happen to the "outside" and a commit can't be pushed to GitHub?

**Reply by surgiie** [reply]
There is no mocking or "sandbox" simulation when running `act`. `act` will run the workflow as it would run in github but in docker containers so it would actually commit and push to github. 

The only difference here is the data that is passed into the workflow so for example

when running the github action on github the `secrets.GH_TOKEN` is populated from the stored secrets in the repository settings but when running in act, i would have to provide this when running the command:

 `act <workflow-name> --secret=GH_TOKEN=<my-token>` 
 
 
 The `skip_fetch` did not work either. Again, my token has plenty of scopes added to it and I am an administrator to the repo in question, so im quite lost here. 😓

**Reply by surgiie** [reply]
As best as i can tell, i think the issue lies with the fact that its trying to push/authenticate with ssh instead of https (via a token)

**Reply by stefanzweifel** [reply]
Thanks for the explanation!

[`actions/checkout`](https://github.com/actions/checkout) handles all authentication with the GitHub repo. According to [this discussion](https://github.com/orgs/community/discussions/50130#discussioncomment-5322500) you could set the `ssh-key` setting to a supported private key so that `ssh` is used to authenticate with GitHub.
By default, checkout should use `https`.

**Reply by surgiie** [reply]
Yeah im a little confused though, according to the checkout action, if https is the default then it should be running git commands with https instead of ssh so its a bit confusing why its not. Since im passing in the token, id expect that to be the case even more:

```
- uses: actions/checkout@master
  with:
    fetch-depth: 2
    token: ${{ secrets.GH_TOKEN }}
```

Since it works in github, im assuming its an issue with `act` and Ill consider opening an issue with  them.

**Reply by surgiie** [reply, marked as answer]
Actually looks like i found my culprit. `act` skips a `actions/checkout` step by copying the local repo into the container. This prevents the `actions/checkout` step from running, therefore preventing the https/config setup to use my token. They have an option to avoid this:

`--no-skip-checkout                                  Do not skip actions/checkout`

**Reply by stefanzweifel** [reply]
@surgiie Glad you found the source of the problem and thank you for documenting the solution here.
I'm sure this will help at least 1other person in the future.

**Reply by ktownsend-cbre** [reply]
This was very helpful in understanding the problem I am having when running with ACT, although for me this action hangs at the following output line instead of failing. Notice the empty branch value.
```
| INPUT_BRANCH value: 
```

I can't use `--no-skip-checkout` because I need it to work from my local directory, but now that I know my issue is caused by skipping the checkout I am working around it with an "if" condition on the step to skip it when running in ACT:
```
if: always() && ${{ !env.ACT }}
```

Now, if I can just figure out how to get ACT to copy the changes back to my local directory so I can review the changes.

**Reply by stefanzweifel** [reply]
@ktownsend-cbre Seeing an empty `INPUT_BRANCH` value is per se not an issue. In this case git-auto-commit would just not switch to a different branch.

The default value for `branch` is `github.head_ref`.
https://github.com/stefanzweifel/git-auto-commit-action/blob/master/action.yml#L11-L14

The code executed after display the INPUT_BRANCH can be found here: https://github.com/stefanzweifel/git-auto-commit-action/blob/0b492c0d951b87f3cd12523a542dbd156c1dbc80/entrypoint.sh#L99-L119

Not sure if this helps resolve your issue. I still haven't used ACT in my projects.

**Reply by ktownsend-cbre** [reply]
@stefanzweifel thanks for clarifying the blank `INPUT_BRANCH` situation. In my case it is actually best if I don't autocommit when running with ACT, so disabling the step is ideal. I got a solution from the ACT project to getting my file changes while running the workflow in ACT reflected on my local. It just requires the `--bind` flag on the ACT command line. If anyone wants the context for that, [go here](https://github.com/nektos/act/discussions/2558).
